### PR TITLE
Additional serial protocol: Graupner HoTT SUMD

### PR DESCRIPTION
**Pull request by mha1**
SUMD is a serial protocol similar to SBUS which is used by some FCs and a variety of flybarless controllers.

In addition Graupner HoTT receivers can be used as a PWM channel extenders or redundant backup system in dual module ELRS/HoTT scenarios.

### CODE REVIEW

File: src/src/rx-serial/SerialSUMD.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
+#include "SerialIO.h"
+
+static const uint16_t crc16_table[256] = {
```
Comment (JyeSmith) [inline src/src/rx-serial/SerialSUMD.h:3]: Instead of adding the table, is the poly known?

Comment (pkendall64) [reply src/src/rx-serial/SerialSUMD.h:3]: The CRC calculation is shown in this document https://www.deviationtx.com/media/kunena/attachments/98/HoTT-SUMD-Spec-REV01-12062012-pdf.pdf on page 6.
It has the polynomial listed there, so the CRC table is not required.

Comment (mha1) [reply src/src/rx-serial/SerialSUMD.h:3]: Yes, I know the spec. I mean it's a tradeoff between speed and flash usage. If you want to save flash I'm happy to put in the polynomial calc. Will update this PR.

**Comment by mha1** [comment]
Anything else I can do?

**Comment by pkendall64** [comment]
> Anything else I can do?

No, it will get reviewed. We're busy getting ready to cut a 3.2.1 release and this change would then be (hopefully) included in 3.3.0 release.

File: src/python/UARTupload.py [hunk 2]
```diff
@@ -257,6 +257,9 @@ def on_upload(source, target, env):
 
     if 3 < len(sys.argv):
         baudrate = sys.argv[3]
+        
+    if 4 < len(sys.argv):
+        baudrate = sys.argv[4]
```
Comment (pkendall64) [inline src/python/UARTupload.py:262]: This change is not required

File: src/python/serve_html.py [hunk 3]
```diff
@@ -61,7 +61,7 @@
                     "pin": 3
                 }
             ],
-            "serial-protocol": 3,
+            "serial-protocol": 4,
```
Comment (pkendall64) [inline src/python/serve_html.py:64]: Not required, I assume this was just for your testing.

File: src/src/rx-serial/SerialSUMD.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#include "SerialIO.h"
+
+#define SUMD_HEADER_SIZE		3														// 3 Bytes header
+#define SUMD_DATA_SIZE_16CH		(16*2)													// 2 Bytes per channel
+#define SUMD_CRC_SIZE			2														// 16 bit CRC
+#define SUMD_FRAME_16CH_LEN		(SUMD_HEADER_SIZE+SUMD_DATA_SIZE_16CH+SUMD_CRC_SIZE)	
+
+#define CRC_POLYNOME 0x1021
```
Comment (pkendall64) [inline src/src/rx-serial/SerialSUMD.h:8]: Move these into the CPP file as they are implementation details.

File: src/python/UARTupload.py (hunk 2 above)
Comment (mha1) [reply src/python/UARTupload.py]: done

File: src/python/serve_html.py (hunk 3 above)
Comment (mha1) [reply src/python/serve_html.py]: done

File: src/src/rx-serial/SerialSUMD.h (hunk 4 above)
Comment (mha1) [reply src/src/rx-serial/SerialSUMD.h]: done

**Comment by mha1** [comment]
@pkendall64 Thanks for your review, please see new commit

File: src/src/rx-serial/SerialSUMD.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	uint16_t crc = crc16(outBuffer, (SUMD_HEADER_SIZE + SUMD_DATA_SIZE_16CH));
+	outBuffer[35] = (uint8_t)(crc >> 8);
+	outBuffer[36] = (uint8_t)(crc & 0x00ff);	  	
+	
+	_outputPort->write(outBuffer, sizeof(outBuffer));
+	
+    return DURATION_IMMEDIATELY;
+}
+
+uint16_t SerialSUMD::crc16(uint8_t *data, uint8_t len)
```
Comment (JyeSmith) [inline src/src/rx-serial/SerialSUMD.cpp:88]: Can this use Crc2Byte to minimise code duplicate?

Sorry I missed this in my first review.

**Comment by JyeSmith** [comment]
Yep, lets get this merged ASAP for v3.3 and some testing.

File: src/src/rx-serial/SerialSUMD.cpp (hunk 5 above)
Comment (mha1) [reply src/src/rx-serial/SerialSUMD.cpp]: Yeah, can do, works.

**Comment by mha1** [comment]
@JyeSmith Please check new commit

**Review by JyeSmith — APPROVED** [review]

**Comment by JyeSmith** [comment]
Looks the good with BF 4.4... cant wait to take it for a rip.

**Comment by mha1** [comment]
@JyeSmith U own Graupner stuff?

**Comment by JyeSmith** [comment]
Nup.

Just a little tongue in check that'll I'll use it with betaflight.

**Comment by mha1** [comment]
have at it

**Review by pkendall64 — APPROVED** [review]
Tested this evening with Betaflight.

**Comment by mha1** [comment]
Thanks guys

**Comment by JyeSmith** [comment]
> Thanks guys

No probs :smile: 

Keep the PRs coming if you have anything else in mind.

**Comment by mha1** [comment]
Actually there's things on my mind:

To be compliant with EU regulations we'd have to implement 868MHz LBT. I say we as I'd need some discussion on how to best do this.

And a completely different thing. As there is a vast number of Graupner HoTT telemetry gear available (Graupner OEM, third party like SM's swiss knife Unisene and DIY stuff like openXsensor) it'd be not too difficult to implement a simple telemetry bus master to poll HoTT telemetry devices and downlink telemetry data via the already ELRS/CRSF telemetry packets. All that's required is (soft)serial half duplex communication on a separate pin, the bus master implementation and repackaging of telemetry data to CRSF packets.

**Comment by JyeSmith** [comment]
@mha1 ping us on discord in the dev chat 👍

**Pull request merged**